src/pages/expansions/StocktakeEditExpansion.js

```diff
@@ -0,0 +1,205 @@
+import { GenericPage } from '../GenericPage';
+import { parsePositiveInteger } from '../../utilities';
+import { tableStrings, buttonStrings } from '../../localization';
+
+/**
+* Renders page to be displayed in StocktakeEditPage -> expansion.
+* @prop   {Realm}               database      App wide database.
+* @state  {Realm.Results}       items         the stocktakeItems of props.stocktake.
+*/
+export class StocktakeEditExpansion extends React.Component {
+  constructor(props) {
+    super(props);
+    this.item = props.data;
+    this.state = {};
+    autobind(this);
+  }
+
+  onEndEditing(key, stocktakeBatch, newValue) {
+    this.props.database.write(() => {
+      switch (key) {
+        case 'countedTotalQuantity': {
+          const newCountedQuantity = parsePositiveInteger(newValue);
+          if (newCountedQuantity === null) return;
+
+          stocktakeBatch.countedTotalQuantity = parsePositiveInteger(newValue);
+          this.props.refreshParent();
+          break;
+        }
+        case 'batch': {
+          if (!newValue || newValue === '') return;
+          stocktakeBatch.batch = newValue;
+          break;
+        }
+        case 'expiryDate':
+          stocktakeBatch.expiryDate = newValue;
+          break;
+        default:
+          break;
+      }
+      this.props.database.save('stocktakeBatch', stocktakeBatch);
+    });
+  }
+
+  refreshData() {
+    this.setState({ data: this.item.batches });
+  }
+
+  renderCell(key, stocktakeBatch) {
+    const isEditable = !this.item.stocktake.isFinalised;
+    switch (key) {
+      default:
+        return {
+          cellContents: stocktakeBatch[key],
+        };
+      case 'batch':
+        return {
+          type: isEditable ? 'editable' : 'text',
+          cellContents: stocktakeBatch[key] && stocktakeBatch[key] !== '' ?
+                        stocktakeBatch[key] : '(no batch name)', // TODO: localise
+        };
+      case 'countedTotalQuantity': {
+        const emptyCellContents = isEditable ? '' : tableStrings.no_change;
+        return {
+          type: isEditable ? 'editable' : 'text',
+          cellContents: stocktakeBatch.difference === 0 ? emptyCellContents :
+                        stocktakeBatch.countedTotalQuantity,
+          placeholder: tableStrings.no_change,
+        };
+      }
+      case 'expiryDate': {
+        return (
+          <ExpiryTextInput
+            key={key}
+            isEditable={isEditable}
+            onEndEditing={(newValue) => {
+              this.onEndEditing(key, stocktakeBatch, newValue);
+              this.refreshData();
+            }}
+            text={stocktakeBatch[key]}
+            style={dataTableStyles.text}
+          />
+        );
+      }
+    }
+  }
+
+  renderAddBatchButton() {
+    const addNewBatch = () => {
+      this.props.database.write(() => {
+        this.item.createNewBatch(this.props.database);
+      });
+      this.refreshData();
+    };
+    return ( // TODO: localise addNewBatch
```

Comment:
Left TODO?

Reply:
I think it's worthwhile "Request changes" a review even if only typos and feedback on code comments.

Reply:
Forgot to remove the TODO: